### PR TITLE
Fix parallel camera access die zweite

### LINKED ISSUE #736: Parallel Camera Initialization Crash
I've noticed that on my robot detection branch the camera crashes sometimes on `image.unwrap()`. (hulk_nao/src/camera.rs:77). This happens only at startup of the framework. This is only fixed by rebooting the NAO (restarting Hulk is not sufficient). Since I only observed this on  my branch, I think this is a bug from #564, probably associated with camera initialization when multiple threads request images

### LINKED ISSUE #961: Camera images sometimes missing at startup
Sometimes at startup we observe the following error:

```
==> hulk/logs/hulk.out <==
2024-04-20 06:42:26  openvino_finder      INFO  Attempting to find library: libopenvino_c.so
2024-04-20 06:42:26  openvino_finder     DEBUG  Searching in: /lib/libopenvino_c.so
2024-04-20 06:42:26  openvino_finder     DEBUG  Searching in: /lib/libopenvino_c.so.2330
2024-04-20 06:42:26  openvino_finder      INFO  Found library at path: /lib/libopenvino_c.so.2330
2024-04-20 06:42:26  openvino_finder      INFO  Attempting to find library: libopenvino_c.so
2024-04-20 06:42:26  openvino_finder     DEBUG  Searching in: /lib/libopenvino_c.so
2024-04-20 06:42:26  openvino_finder     DEBUG  Searching in: /lib/libopenvino_c.so.2330
2024-04-20 06:42:26  openvino_finder      INFO  Found library at path: /lib/libopenvino_c.so.2330
2024-04-20 06:42:26  openvino_finder     DEBUG  Searching in: /lib/plugins.xml
2024-04-20 06:42:26  openvino_finder     DEBUG  Searching in: /lib/openvino-2023.3.0/plugins.xml

==> hulk/logs/hulk.err <==
panicked at crates/hulk_nao/src/camera.rs:77:18:
called `Option::unwrap()` on a `None` value
failed to join cycler thread ObjectDetectionTop: Any { .. }
Error: 
   0: at least one cycler exited with error

Location:
   /home/h3ndrk/Documents/hulks/hulk3/target/x86_64-aldebaran-linux-gnu/incremental/build/hulk-3260f5f5b832c76b/out/generated_code.rs:5425

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

### PR DESCRIPTION
**Pull request by oleflb**
## Introduced Changes

Fixes the camera access bug. Also some minor refactoring in the camera, to remove one layer of interior mutability.

Fixes #961 
Fixes #736

## ToDo / Known Issues

*If this is a WIP describe which problems are to be fixed.*

## Ideas for Next Iterations (Not This PR)

*If there are some improvements that could be done in a next iteration, describe them here.*

## How to Test

- Upload to NAO (should never crash)
- check vision recordings, should never process same frame twice

### CODE REVIEW

File: crates/hulk_nao/src/camera.rs
Comment (h3ndrk) [inline crates/hulk_nao/src/camera.rs:1]: Could you please document somewhere what the intended semantics, i.e., the contract, are s.t. we can review and check it against that?

Comment (oleflb) [reply crates/hulk_nao/src/camera.rs:1]: Done

File: crates/hulk_nao/src/camera.rs [hunk 1]
```diff
@@ -3,25 +3,30 @@ use std::{
...
-use parking_lot::{Mutex, RwLock};
+use parking_lot::Mutex;
 use types::{camera_position::CameraPosition, ycbcr422_image::YCbCr422Image};
+use watch::WatchSender as Sender;
 
 pub struct Camera {
-    camera: RwLock<Option<NaoCamera>>,
-    path: PathBuf,
+    camera: Mutex<CameraHardware>,
+    image_sender: Sender<Option<YCbCr422Image>>,
```
Comment (h3ndrk) [inline crates/hulk_nao/src/camera.rs:18]: Why is the item type `Option`? Did I overlook the case where you send a `None`? Can we remove the `Option`?

File: crates/hulk_nao/src/camera.rs [hunk 2]
```diff
@@ -31,60 +36,73 @@ impl Camera {
...
+
+impl CameraHardware {
+    fn get_next_image(&mut self) -> Result<YCbCr422Image> {
+        self.wait_for_device()
+            .wrap_err("failed to wait for device")?;
 
-    fn wait_for_device(&self) -> Result<()> {
+        let Some(camera) = self.camera.as_mut() else {
+            bail!("camera does not exist")
+        };
```
Comment (h3ndrk) [inline crates/hulk_nao/src/camera.rs:80]: Isn't there a monadic way of expressing this `let Some else`? Like `unwrap_or` or `or_else`. I remember some method for converting a `None` into a `Err(...)`. 🤔

```suggestion
        let camera = self.camera.as_mut().ok_or_else(|| eyre!("camera does not exist"))?;
```

File: crates/hulk_nao/src/camera.rs [hunk 3]
```diff
@@ -3,25 +3,30 @@ use std::{
...
 use nao_camera::{reset_camera_device, Camera as NaoCamera, Parameters, PollingError};
-use parking_lot::{Mutex, RwLock};
+use parking_lot::Mutex;
 use types::{camera_position::CameraPosition, ycbcr422_image::YCbCr422Image};
+use watch::WatchSender as Sender;
 
 pub struct Camera {
-    camera: RwLock<Option<NaoCamera>>,
-    path: PathBuf,
+    camera: Mutex<CameraHardware>,
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/hulk_nao/src/camera.rs:17]: Talked about this, it is one mutex per camera.

File: crates/hulk_nao/src/camera.rs (hunk 1 above)
Comment (h3ndrk) [reply crates/hulk_nao/src/camera.rs:18]: The channel needs to be initialized with something that is defaultable.

**Review by h3ndrk — CHANGES_REQUESTED** [review]

**Review by h3ndrk — APPROVED** [review]

**Pull request merged**